### PR TITLE
Feature: Filter component

### LINKED ISSUE #2218: Filter Component
## 🗒 General

Would love some help building out this filter component for Storybook! Includes several different input options, including single select dropdown, multiselect dropdown, and nested subcategories. This filter is meant to be used with a full page [Table ](https://backstage.io/storybook/?path=/story/table--default-table)component.

Filters hidden
Filters revealed

## 💻 Usage
Usage details can be found in the screenshots below.

Filters revealed _ Expanded Dropdown _ Many Checkbox Chips
Filters revealed _ Expanded Dropdown _ Checkbox Chips
Filters revealed _ Expanded Dropdown _ Checkbox
Filters revealed _ Expanded Dropdown
Filters revealed

## 📐 Specs
You can access specs by duplicating our [Figma UI kit](https://www.figma.com/community/file/850673348101741100). Click on the 'Filters' page on the left hand side. 
Screenshot 2020-09-01 at 5 59 17 PM

### PR DESCRIPTION
**Pull request by ayshiff**
## Hey, I just made a Pull Request!

Closes #2218

This is still a work in progress because I need to do some **adjustments (code + style), refactoring and add tests**. (There is still a lot of messy code)

This PR introduces two new types of components: the `Select` and the `CheckboxTree`.

- `Select`
    - Default
Capture d’écran 2020-09-15 à 18 29 35

    - Multiple
Capture d’écran 2020-09-15 à 18 29 59

- `CheckboxTree`
Capture d’écran 2020-09-15 à 18 29 50

It also introduce the `Filter` component inside the `Table`.
Capture d’écran 2020-09-15 à 18 30 15

Here is **how it works** [**Outdated**]:
filter-table

## Notes and remarks

Things left to do:
- [x] Add missing tests
- [ ] Add search bar to the multiple select
- [x] Clear all: resets the data without filter but I need to reset the filters initial states
- [ ] Add missing styles
- [ ] Global refactoring

I have some questions about some **UX/UI behaviors**:
- What is the behavior of the filter tab when the number of filters exceeds the size (height) of the table ? Maybe a scroll ?
- I don't know in which case the `CheckboxTree` should have 2 levels ? So I can reproduce a working example.
- Do the filters have to add up?
What I mean is if we have checked `value1` for the column1 and `value2` for the column2 do we have to display the data with the column1 = `value1` and column2 = `value2` ?
(For the moment the logic cover this type of behavior)
[**EDIT**: as @freben said below, we have an AND logic between different groups of filters and an OR logic inside the same group of filter.]

Moreover I find my implementation of the `CheckboxTree` component a bit complex.
For the items `props` we have something like:

```json
[
   {
      "label":"Genereic subcategory name 1",
      "options":[
         {
            "label":"Option 1",
            "value":1
         },
         {
            "label":"Option 2",
            "value":2
         }
      ]
   }
]
```

And then in the component, for performance purpose we need to index each item to obtain something like this:
```json
{
   "Genereic subcategory name 1":{
      "label":"Genereic subcategory name 1",
      "isOpen":false,
      "isChecked":false,
      "options":{
         "Option 1":{
            "label":"Option 1",
            "isChecked":false,
            "value":1
         },
         "Option 2":{
            "label":"Option 2",
            "isChecked":false,
            "value":2
         }
      }
   }
}
```

And we have a **reducer with 3 types of actions**: `checkOption`, `checkCategory` and `openCategory` (that should be named `toggleCategory` [**EDIT**: fixed]) that mutate our state which is our only source of truth.
Maybe we should **split the responsibilities** (a stage managing the checked options, another managing the opening categories...) to have a cleaner codebase.

But I like having an object that contains all of our properties. (This way **we keep the notion of hierarchy between categories and options**). 
The problem with this method is that for every change I have to go deep through my object to change a specific value.

[**EDIT**: fixed by using the [immer](https://immerjs.github.io/immer/docs/produce) `produce` function]

Can I have your point of view on this implementation ?

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [x] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [x] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

**Comment by katz95** [comment]
😍 😍 😍

**Comment by freben** [comment]
I definitely expected the groups of filters to be having an AND between them (which they do not seem to have in your gif), and OR within each group (which seems to match your gif). E.g. in the gif when you select "More data 1", I expect the result set to be empty because there are no entries that match both that, and either of 141 / 249. This is how I feel that filters generally work on e-commerce sites and similar.

File: packages/core/src/components/CheckboxTree/CheckboxTree.tsx [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        ...state,
+        [action.payload.subCategoryLabel]: {
+          ...state[action.payload.subCategoryLabel],
+          isChecked: Object.values(newOptions).every(
+            option => option.isChecked,
+          ),
+          options: newOptions,
+        },
+      };
+    }
```
Comment (freben) [inline packages/core/src/components/CheckboxTree/CheckboxTree.tsx:154]: Reducers in plain js tend to get ultra gnarly. Would something like immer help you condense these into much more readable code? All of the spreads will go away.

https://immerjs.github.io/immer/docs/introduction

```ts
return produce(state, newState => {
  const category = newState[action.payload.subCategoryLabel];
  const option = category.options[action.payload.optionLabel];
  option.selected = !option.selected;
  category.selected = category.options.every(o => o.selected);
});
```

**Comment by soapraj** [comment]
Hey, @ayshiff is this good to be promoted from a draft -> PR?

I went through the [UI storybook tests](https://www.chromatic.com/build?appId=5eafd13eaebb8f00227540f4&number=1231) in chromatic and in some cases, the tables width have become smaller - can you take a look

**Comment by ayshiff** [comment]
Hey @soapraj !
I was looking to add the search bar to the multiple select and cover some edge cases (UI/UX behaviors) that I mentionned in the description:

*- What is the behavior of the filter tab when the number of filters exceeds the size (height) of the table ? Maybe a scroll ?*
*- I don't know in which case the CheckboxTree should have 2 levels ? So I can reproduce a working example.*

But I think I will promote it to a PR as we can cover these edge cases in a followup PR (As this one starts to get a little big)
Yes you are right, I'll fix that 👍

**Comment by shmidt-i** [comment]
Hey @ayshiff, thanks for this contribution, much appreciated! 
I think it's better to merge it sooner than later and iterate on it! Code looks pretty good to me, found a small bug, but again - maybe fix it in a separate PR: in "multiple" select story there is possibility to click the "All results", which is just a placeholder and it gets you to this weird state - 
Screenshot 2020-09-24 at 11 36 30

**Comment by ayshiff** [comment]
I fixed the problem that @shmidt-i reported in the multiple select.
If you are happy with the code, this PR should be ready to merge 👍

**Review by benjdlambert — COMMENTED** [review]
Let's get this in and create tickets for anything else we need to do after.

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**